### PR TITLE
Add network support to Pi2/3

### PR DESCRIPTION
**Pull request by hardillb**
Also include most of the changes needed for Qemu 5.1 from #20 as needed for USB support (with the resize increased to 4G and moved to apply to passed in images as well as the included one).

Should have the included image updated to match latest at time of release.

Device comes up as usb0 but close enough to get started.

I did test with Qemu 5.2-rc3 to see if Pi4 support was included but doesn't look like it yet.

### CODE REVIEW

**Comment by hardillb** [comment]
I've moved the resize to only be applied for Pi3 (64bit builds) as it's only needed there and it now resizes the image to the closest next 2G increment rather than fixed at 4G (the lite image is good at 2G)

**Comment by davystrong** [comment]
Any chance of this getting merged?

**Comment by jeremypoulter** [comment]
For what it is worth I tried this and can confirm it is working

**Comment by lukechilds** [comment]
Thanks for the review @jeremypoulter and thanks again for working on this @hardillb. Apologies for how long it's taken for me to review, I'm really busy with @getumbrel at the moment. No promises, but I've put a note in my calendar to try and take the time to review, test and merge this at the weekend.

File: Dockerfile [hunk 1]
```diff
@@ -89,8 +90,8 @@ ENTRYPOINT ["./entrypoint.sh"]
 # It's just the VM image with a compressed Raspbian filesystem added
 FROM dockerpi-vm as dockerpi
 LABEL maintainer="Luke Childs <lukechilds123@gmail.com>"
-ARG FILESYSTEM_IMAGE_URL="http://downloads.raspberrypi.org/raspbian_lite/images/raspbian_lite-2019-09-30/2019-09-26-raspbian-buster-lite.zip"
-ARG FILESYSTEM_IMAGE_CHECKSUM="a50237c2f718bd8d806b96df5b9d2174ce8b789eda1f03434ed2213bbca6c6ff"
+ARG FILESYSTEM_IMAGE_URL="https://downloads.raspberrypi.org/raspios_lite_armhf/images/raspios_lite_armhf-2021-03-25/2021-03-04-raspios-buster-armhf-lite.zip"
```
Comment (lukechilds) [inline Dockerfile:93]: Can you revert this change?

We can add it in a separate PR.

File: entrypoint.sh [hunk 2]
```diff
@@ -15,28 +15,38 @@ if [ ! -e $image_path ]; then
...
   dtb="/root/qemu-rpi-kernel/versatile-pb.dtb"
   machine=versatilepb
   memory=256m
   root=/dev/sda2
+  extra=''
   nic='--net nic --net user,hostfwd=tcp::5022-:22'
 elif [ "${target}" = "pi2" ]; then
   emulator=qemu-system-arm
-  machine=raspi2
+  machine=raspi2b
```
Comment (lukechilds) [inline entrypoint.sh:36]: Why did you change from `raspi2` to `raspi2b` here?

File: entrypoint.sh [hunk 3]
```diff
@@ -15,28 +15,38 @@ if [ ! -e $image_path ]; then
   fi
 fi
 
+echo "Rounding image size up to a multiple of 2G"
+image_size=`du -m $image_path | cut -f1`
```
Comment (lukechilds) [inline entrypoint.sh:19]: Are you sure this calculation is correct?

If I run:

```
docker run -it -v /tmp/dockerpi:/sdcard dockerpi
```

Then the image gets persisted at `/tmp/dockerpi/filesystem.img`.

However dockerpi attempts to resize it on every boot saying:

```
Rounding image size up to a multiple of 2G
from 2145M to 4G
```

This command seems to disagree with `ls`:

```
$ du -m /tmp/dockerpi/filesystem.img | cut -f1
2145

$ ls -lah /tmp/dockerpi/filesystem.img
-rw-r--r--@ 1 lukechilds  wheel   4.0G 16 May 15:45 /tmp/dockerpi/filesystem.img
```

If the image is already an appropriate size, we shouldn't attempt to resize it again.

File: Dockerfile [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Build stage for qemu-system-arm
 FROM debian:stable-slim AS qemu-builder
-ARG QEMU_VERSION=4.2.0
+ARG QEMU_VERSION=5.2.0
```
Comment (lukechilds) [inline Dockerfile:3]:
```suggestion
ARG QEMU_VERSION=6.0.0
```

QEMU 6.0.0 is out now. Lets update to that.

File: Dockerfile [hunk 5]
```diff
@@ -67,6 +67,7 @@ ARG RPI_KERNEL_CHECKSUM="295a22f1cd49ab51b9e7192103ee7c917624b063cc5ca2e11434164
 
 COPY --from=qemu-builder /qemu/arm-softmmu/qemu-system-arm /usr/local/bin/qemu-system-arm
 COPY --from=qemu-builder /qemu/aarch64-softmmu/qemu-system-aarch64 /usr/local/bin/qemu-system-aarch64
+COPY --from=qemu-builder /qemu/qemu-img /usr/local/bin/qemu-img
```
Comment (lukechilds) [inline Dockerfile:70]: We should `strip` this binary in the builder image first.

**Review by lukechilds — CHANGES_REQUESTED** [review]
Thanks for this awesome PR!

Some small changes.

File: entrypoint.sh (hunk 2 above)
Comment (hardillb) [reply entrypoint.sh:36]: Because it matches the published name in qemu, `raspi2` is an alias for `raspib2`

File: Dockerfile (hunk 5 above)
Comment (lukechilds) [reply Dockerfile:70]: This reduces the binary from 15.6M to  3.4M!

File: Dockerfile (hunk 1 above)
Comment (hardillb) [reply Dockerfile:93]: OK

File: Dockerfile (hunk 5 above)
Comment (lukechilds) [reply Dockerfile:70]: Resolved via https://github.com/lukechilds/dockerpi/pull/21/commits/9809ecd0709f31de09d3cd4cbeb2cffefb29f477

File: Dockerfile (hunk 1 above)
Comment (hardillb) [reply Dockerfile]: Should have the old version now (with the new one commented out to make it easier later)

Comment (lukechilds) [reply Dockerfile]: Thanks, I just meant remove it completely so there's no noise in the squashed commit. All good, resolved in https://github.com/lukechilds/dockerpi/pull/21/commits/767fa33c0d4c393178777f02b7068d17e0c98500.

File: entrypoint.sh (hunk 3 above)
Comment (hardillb) [reply entrypoint.sh:19]: The resize is near instant, it's just a change to the filesize so the image become a sparse file. So it shouldn't even consume any more storage. But if you can think of a better way to do it.

Comment (lukechilds) [reply entrypoint.sh]: Sooo after a some investigating it looks like the original code was checking size on disk but what QEMU is checking is the virtual size:

```
$ qemu-img info /sdcard/filesystem.img
image: /sdcard/filesystem.img
file format: raw
virtual size: 4 GiB (4294967296 bytes)
disk size: 2.09 GiB
```

Resolved in https://github.com/lukechilds/dockerpi/pull/21/commits/5020c6394f49598b4fefa266b39ed34f79841504.

**Review by lukechilds — APPROVED** [review]
Thanks so much for your help with this @hardillb!

This looks all good to me now, are you able to give my latest changes a quick eyeball and let me know if anything looks off to you?

**Comment by hardillb** [comment]
I'm out and about for a few hours, I can give it a proper look when i get back home

**Comment by hardillb** [comment]
All looks good

The only thing I noticed, it looks like the pi3 option will only work with the 64bit RaspiOS builds not the 32bit builds because it's using the kernel8.img but looking at the history this would have been the case before the merge so not a regression.

**Comment by lukechilds** [comment]
I think the current images are 32bit, they're working ok for me.

**Comment by lukechilds** [comment]
kernel8.img is included in the image so wouldn't this be 32 bit or 64 bit depending on which image is being used?

**Comment by hardillb** [comment]
You can get 64bit images here https://downloads.raspberrypi.org/raspios_arm64/images/ they are just not listed on the download pages yet.

Running the vm image against a 32bit image hangs and maxes out the CPUs for me.
e.g.
```
docker run --name dockerpi -it --rm -v `pwd`/2021-03-04-raspios-buster-armhf-lite.img:/sdcard/filesystem.img ben/dockerpi:vm pi3
```
Hangs here:
```
...
Searching for kernel='kernel8.img'
Searching for dtb='bcm2710-rpi-3-b-plus.dtb'
Booting QEMU machine "raspi3b" with kernel=/fat/kernel8.img dtb=/fat/bcm2710-rpi-3-b-plus.dtb
```

**Comment by lukechilds** [comment]
Strange, 32 bit works fine for me: [2021-03-04-raspios-buster-armhf-lite.zip](https://downloads.raspberrypi.org/raspios_lite_armhf/images/raspios_lite_armhf-2021-03-25/2021-03-04-raspios-buster-armhf-lite.zip)

But I get a kernel panic (no hanging) with 64 bit: [2021-03-04-raspios-buster-arm64.zip](https://downloads.raspberrypi.org/raspios_arm64/images/raspios_arm64-2021-04-09/2021-03-04-raspios-buster-arm64.zip).

Ah well, like you said, this is already the current behaviour, so I'll go ahead and merge this PR. Thanks again for your help!

**Review by lukechilds — APPROVED** [review]
